src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -716,6 +716,10 @@ private function addNewInstance(Definition $definition, $return, $instantiation)
                 $class = $this->dumpValue($callable[0]);
                 // If the class is a string we can optimize call_user_func away
                 if (strpos($class, "'") === 0) {
+                    if ("''" === $class) {
+                        throw new RuntimeException(sprintf('The "%s" service is defined to be created by a factory but is missing the service reference, did you forget to define the factory service id or class?', $id));
```

Comment:
updated the message, added the mention of class to what @iltar proposed (you can specify a factory by class and static method instead of service id)